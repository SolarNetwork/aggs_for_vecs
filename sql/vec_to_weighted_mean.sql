-- smallint weighted_mean lots
SELECT vec_to_weighted_mean(smallints, smallints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- smallint weighted_mean none
SELECT vec_to_weighted_mean(smallints, smallints) FROM measurements WHERE sensor_id = -1;
-- smallint weighted_mean one null
SELECT vec_to_weighted_mean(smallints, smallints) FROM measurements WHERE sensor_id = 1;
-- smallint weighted_mean array of nulls
SELECT vec_to_weighted_mean(smallints, smallints) FROM measurements WHERE sensor_id = 2;
-- smallint weighted_mean one not-null
SELECT vec_to_weighted_mean(smallints, smallints) FROM measurements WHERE sensor_id = 4;
-- smallint weighted_mean array of nulls and one other
SELECT vec_to_weighted_mean(smallints, smallints) FROM measurements WHERE sensor_id IN (2, 4);

-- int weighted_mean lots
SELECT vec_to_weighted_mean(ints, ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- int weighted_mean none
SELECT vec_to_weighted_mean(ints, ints) FROM measurements WHERE sensor_id = -1;
-- int weighted_mean one null
SELECT vec_to_weighted_mean(ints, ints) FROM measurements WHERE sensor_id = 1;
-- int weighted_mean array of nulls
SELECT vec_to_weighted_mean(ints, ints) FROM measurements WHERE sensor_id = 2;
-- int weighted_mean one not-null
SELECT vec_to_weighted_mean(ints, ints) FROM measurements WHERE sensor_id = 4;
-- int weighted_mean array of nulls and one other
SELECT vec_to_weighted_mean(ints, ints) FROM measurements WHERE sensor_id IN (2, 4);

-- bigint weighted_mean lots
SELECT vec_to_weighted_mean(bigints, bigints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- bigint weighted_mean none
SELECT vec_to_weighted_mean(bigints, bigints) FROM measurements WHERE sensor_id = -1;
-- bigint weighted_mean one null
SELECT vec_to_weighted_mean(bigints, bigints) FROM measurements WHERE sensor_id = 1;
-- bigint weighted_mean array of nulls
SELECT vec_to_weighted_mean(bigints, bigints) FROM measurements WHERE sensor_id = 2;
-- bigint weighted_mean one not-null
SELECT vec_to_weighted_mean(bigints, bigints) FROM measurements WHERE sensor_id = 4;
-- bigint weighted_mean array of nulls and one other
SELECT vec_to_weighted_mean(bigints, bigints) FROM measurements WHERE sensor_id IN (2, 4);

-- real weighted_mean lots
SELECT vec_to_weighted_mean(reals, reals) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- real weighted_mean none
SELECT vec_to_weighted_mean(reals, reals) FROM measurements WHERE sensor_id = -1;
-- real weighted_mean one null
SELECT vec_to_weighted_mean(reals, reals) FROM measurements WHERE sensor_id = 1;
-- real weighted_mean array of nulls
SELECT vec_to_weighted_mean(reals, reals) FROM measurements WHERE sensor_id = 2;
-- real weighted_mean one not-null
SELECT vec_to_weighted_mean(reals, reals) FROM measurements WHERE sensor_id = 4;
-- real weighted_mean array of nulls and one other
SELECT vec_to_weighted_mean(reals, reals) FROM measurements WHERE sensor_id IN (2, 4);

-- float weighted_mean lots
SELECT vec_to_weighted_mean(floats, floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- float weighted_mean none
SELECT vec_to_weighted_mean(floats, floats) FROM measurements WHERE sensor_id = -1;
-- float weighted_mean one null
SELECT vec_to_weighted_mean(floats, floats) FROM measurements WHERE sensor_id = 1;
-- float weighted_mean array of nulls
SELECT vec_to_weighted_mean(floats, floats) FROM measurements WHERE sensor_id = 2;
-- float weighted_mean one not-null
SELECT vec_to_weighted_mean(floats, floats) FROM measurements WHERE sensor_id = 4;
-- float weighted_mean array of nulls and one other
SELECT vec_to_weighted_mean(floats, floats) FROM measurements WHERE sensor_id IN (2, 4);
