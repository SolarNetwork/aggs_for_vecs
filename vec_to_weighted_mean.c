Datum vec_to_weighted_mean_transfn(PG_FUNCTION_ARGS);
PG_FUNCTION_INFO_V1(vec_to_weighted_mean_transfn);

/**
 * Returns an of n elements,
 * which each element is the weighted mean of the values found in that position
 * from all input arrays. The second parameter is an array of weights.
 *
 * by Paul A. Jungwirth
 */
Datum
vec_to_weighted_mean_transfn(PG_FUNCTION_ARGS)
{
  Oid elemTypeId;
  Oid elemWeightTypeId;
  int16 elemTypeWidth;
  int16 elemWeightTypeWidth;
  bool elemTypeByValue;
  bool elemWeightTypeByValue;
  char elemTypeAlignmentCode;
  char elemWeightTypeAlignmentCode;
  int currentLength;
  int currentWeightLength;
  MemoryContext aggContext;
  VecArrayBuildState *state = NULL;
  ArrayType *currentArray;
  ArrayType *currentWeightArray;
  int arrayLength;
  int arrayLengthWeight;
  Datum *currentVals;
  Datum *currentWeightVals;
  bool *currentNulls;
  bool *currentWeightNulls;
  int i;
  float8 prevValue, prevWeight, currentValue, currentWeight, newWeight;

  if (!AggCheckCallContext(fcinfo, &aggContext)) {
    elog(ERROR, "vec_to_weighted_mean_transfn called in non-aggregate context");
  }

  // PG_ARGISNULL tests for SQL NULL,
  // but after the first pass we can have a
  // value that is non-SQL-NULL but still is C NULL.
  if (!PG_ARGISNULL(0)) {
    state = (VecArrayBuildState *)PG_GETARG_POINTER(0);
  }

  if (PG_ARGISNULL(1) || PG_ARGISNULL(2)) {
    // just return the current state unchanged (possibly still NULL)
    PG_RETURN_POINTER(state);
  }
  currentArray = PG_GETARG_ARRAYTYPE_P(1);
  currentWeightArray = PG_GETARG_ARRAYTYPE_P(2);

  if (state == NULL) {
    // Since we have our first not-null argument
    // we can initialize the state to match its length.
    elemTypeId = ARR_ELEMTYPE(currentArray);
    if (elemTypeId != INT2OID &&
        elemTypeId != INT4OID &&
        elemTypeId != INT8OID &&
        elemTypeId != FLOAT4OID &&
        elemTypeId != FLOAT8OID) {
      ereport(ERROR, (errmsg("vec_to_weighted_mean first input must be array of SMALLINT, INTEGER, BIGINT, REAL, or DOUBLE PRECISION")));
    }
    elemWeightTypeId = ARR_ELEMTYPE(currentWeightArray);
    if (elemWeightTypeId != INT2OID &&
        elemWeightTypeId != INT4OID &&
        elemWeightTypeId != INT8OID &&
        elemWeightTypeId != FLOAT4OID &&
        elemWeightTypeId != FLOAT8OID) {
      ereport(ERROR, (errmsg("vec_to_weighted_mean weights input must be array of SMALLINT, INTEGER, BIGINT, REAL, or DOUBLE PRECISION")));
    }
    if (ARR_NDIM(currentArray) != 1) {
      ereport(ERROR, (errmsg("One-dimensional arrays are required")));
    }
    if (ARR_NDIM(currentWeightArray) != 1) {
      ereport(ERROR, (errmsg("One-dimensional arrays are required for weights")));
    }
    arrayLength = (ARR_DIMS(currentArray))[0];
    arrayLengthWeight = (ARR_DIMS(currentWeightArray))[0];
    if (arrayLength != arrayLengthWeight) {
      ereport(ERROR, (errmsg("All arrays must be the same length, but we got %d for values vs %d for weights", arrayLength, arrayLengthWeight)));
    }
    // Just start with all NULLs and let the comparisons below replace them:
    state = initVecArrayResultWithNulls(elemTypeId, FLOAT8OID, aggContext, arrayLength);
  } else {
    elemTypeId = state->inputElementType;
    arrayLength = state->state.nelems;
    elemWeightTypeId = ARR_ELEMTYPE(currentWeightArray);
  }

  get_typlenbyvalalign(elemTypeId, &elemTypeWidth, &elemTypeByValue, &elemTypeAlignmentCode);
  deconstruct_array(currentArray, elemTypeId, elemTypeWidth, elemTypeByValue, elemTypeAlignmentCode,
      &currentVals, &currentNulls, &currentLength);
  if (currentLength != arrayLength) {
    ereport(ERROR, (errmsg("All arrays must be the same length, but we got %d vs %d", currentLength, arrayLength)));
  }

  get_typlenbyvalalign(elemWeightTypeId, &elemWeightTypeWidth, &elemWeightTypeByValue, &elemWeightTypeAlignmentCode);
  deconstruct_array(currentWeightArray, elemWeightTypeId, elemWeightTypeWidth, elemWeightTypeByValue, elemWeightTypeAlignmentCode,
      &currentWeightVals, &currentWeightNulls, &currentWeightLength);
  if (currentWeightLength != arrayLength) {
    ereport(ERROR, (errmsg("All arrays must be the same length, but we got %d vs %d for weights", arrayLength, currentLength)));
  }

  for (i = 0; i < arrayLength; i++) {
    if (currentNulls[i] || currentWeightNulls[i]) {
      // do nothing: nulls can't change the result.
    } else {
      switch (elemTypeId) {
        case INT2OID:   currentValue = DatumGetInt16(currentVals[i]);  break;
        case INT4OID:   currentValue = DatumGetInt32(currentVals[i]);  break;
        case INT8OID:   currentValue = DatumGetInt64(currentVals[i]);  break;
        case FLOAT4OID: currentValue = DatumGetFloat4(currentVals[i]); break;
        case FLOAT8OID: currentValue = DatumGetFloat8(currentVals[i]); break;
        default: elog(ERROR, "Unknown elemTypeId!");
      }
      switch (elemWeightTypeId) {
        case INT2OID:   currentWeight = DatumGetInt16(currentWeightVals[i]);  break;
        case INT4OID:   currentWeight = DatumGetInt32(currentWeightVals[i]);  break;
        case INT8OID:   currentWeight = DatumGetInt64(currentWeightVals[i]);  break;
        case FLOAT4OID: currentWeight = DatumGetFloat4(currentWeightVals[i]); break;
        case FLOAT8OID: currentWeight = DatumGetFloat8(currentWeightVals[i]); break;
        default: elog(ERROR, "Unknown elemTypeId!");
      }
      if (state->state.dnulls[i]) {
        state->state.dnulls[i] = false;
        state->vecvalues[i].f8 = currentValue;
        state->vectmpvalues[i].f8 = currentWeight;
      } else {
        prevValue = state->vecvalues[i].f8;
        prevWeight = state->vectmpvalues[i].f8;
        newWeight = prevWeight + currentWeight;
        if (!newWeight) {
          state->vecvalues[i].f8 = 0;
          state->vectmpvalues[i].f8 = 0;
        } else {
          state->vecvalues[i].f8 = (prevWeight * prevValue + currentWeight * currentValue) / newWeight;
          state->vectmpvalues[i].f8 = newWeight;
        }
      }
    }
  }
  PG_RETURN_POINTER(state);
}

Datum vec_to_weighted_mean_finalfn(PG_FUNCTION_ARGS);
PG_FUNCTION_INFO_V1(vec_to_weighted_mean_finalfn);

Datum
vec_to_weighted_mean_finalfn(PG_FUNCTION_ARGS)
{
  Datum result;
  VecArrayBuildState *state;
  int dims[1];
  int lbs[1];
  int i;

  Assert(AggCheckCallContext(fcinfo, NULL));

  state = PG_ARGISNULL(0) ? NULL : (VecArrayBuildState *)PG_GETARG_POINTER(0);

  if (state == NULL)
    PG_RETURN_NULL();

  // Convert from our pgnums to Datums:
  for (i = 0; i < state->state.nelems; i++) {
    if (state->state.dnulls[i]) continue;
    state->state.dvalues[i] = Float8GetDatum(state->vecvalues[i].f8);
  }

  dims[0] = state->state.nelems;
  lbs[0] = 1;

  result = makeMdArrayResult(&state->state, 1, dims, lbs, CurrentMemoryContext, false);
  PG_RETURN_DATUM(result);
}
