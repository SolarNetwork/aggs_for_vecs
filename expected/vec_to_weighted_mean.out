-- smallint weighted_mean lots
SELECT vec_to_weighted_mean(smallints, smallints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_weighted_mean 
----------------------
 {1,2,2.6}
(1 row)

-- smallint weighted_mean none
SELECT vec_to_weighted_mean(smallints, smallints) FROM measurements WHERE sensor_id = -1;
 vec_to_weighted_mean 
----------------------
 
(1 row)

-- smallint weighted_mean one null
SELECT vec_to_weighted_mean(smallints, smallints) FROM measurements WHERE sensor_id = 1;
 vec_to_weighted_mean 
----------------------
 
(1 row)

-- smallint weighted_mean array of nulls
SELECT vec_to_weighted_mean(smallints, smallints) FROM measurements WHERE sensor_id = 2;
 vec_to_weighted_mean 
----------------------
 {NULL,NULL,NULL}
(1 row)

-- smallint weighted_mean one not-null
SELECT vec_to_weighted_mean(smallints, smallints) FROM measurements WHERE sensor_id = 4;
 vec_to_weighted_mean 
----------------------
 {1,NULL,2}
(1 row)

-- smallint weighted_mean array of nulls and one other
SELECT vec_to_weighted_mean(smallints, smallints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_weighted_mean 
----------------------
 {1,NULL,2}
(1 row)

-- int weighted_mean lots
SELECT vec_to_weighted_mean(ints, ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_weighted_mean 
----------------------
 {1,2,2.6}
(1 row)

-- int weighted_mean none
SELECT vec_to_weighted_mean(ints, ints) FROM measurements WHERE sensor_id = -1;
 vec_to_weighted_mean 
----------------------
 
(1 row)

-- int weighted_mean one null
SELECT vec_to_weighted_mean(ints, ints) FROM measurements WHERE sensor_id = 1;
 vec_to_weighted_mean 
----------------------
 
(1 row)

-- int weighted_mean array of nulls
SELECT vec_to_weighted_mean(ints, ints) FROM measurements WHERE sensor_id = 2;
 vec_to_weighted_mean 
----------------------
 {NULL,NULL,NULL}
(1 row)

-- int weighted_mean one not-null
SELECT vec_to_weighted_mean(ints, ints) FROM measurements WHERE sensor_id = 4;
 vec_to_weighted_mean 
----------------------
 {1,NULL,2}
(1 row)

-- int weighted_mean array of nulls and one other
SELECT vec_to_weighted_mean(ints, ints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_weighted_mean 
----------------------
 {1,NULL,2}
(1 row)

-- bigint weighted_mean lots
SELECT vec_to_weighted_mean(bigints, bigints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_weighted_mean 
----------------------
 {1,2,2.6}
(1 row)

-- bigint weighted_mean none
SELECT vec_to_weighted_mean(bigints, bigints) FROM measurements WHERE sensor_id = -1;
 vec_to_weighted_mean 
----------------------
 
(1 row)

-- bigint weighted_mean one null
SELECT vec_to_weighted_mean(bigints, bigints) FROM measurements WHERE sensor_id = 1;
 vec_to_weighted_mean 
----------------------
 
(1 row)

-- bigint weighted_mean array of nulls
SELECT vec_to_weighted_mean(bigints, bigints) FROM measurements WHERE sensor_id = 2;
 vec_to_weighted_mean 
----------------------
 {NULL,NULL,NULL}
(1 row)

-- bigint weighted_mean one not-null
SELECT vec_to_weighted_mean(bigints, bigints) FROM measurements WHERE sensor_id = 4;
 vec_to_weighted_mean 
----------------------
 {1,NULL,2}
(1 row)

-- bigint weighted_mean array of nulls and one other
SELECT vec_to_weighted_mean(bigints, bigints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_weighted_mean 
----------------------
 {1,NULL,2}
(1 row)

-- real weighted_mean lots
SELECT vec_to_weighted_mean(reals, reals) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_weighted_mean 
----------------------
 {1,2,2.6}
(1 row)

-- real weighted_mean none
SELECT vec_to_weighted_mean(reals, reals) FROM measurements WHERE sensor_id = -1;
 vec_to_weighted_mean 
----------------------
 
(1 row)

-- real weighted_mean one null
SELECT vec_to_weighted_mean(reals, reals) FROM measurements WHERE sensor_id = 1;
 vec_to_weighted_mean 
----------------------
 
(1 row)

-- real weighted_mean array of nulls
SELECT vec_to_weighted_mean(reals, reals) FROM measurements WHERE sensor_id = 2;
 vec_to_weighted_mean 
----------------------
 {NULL,NULL,NULL}
(1 row)

-- real weighted_mean one not-null
SELECT vec_to_weighted_mean(reals, reals) FROM measurements WHERE sensor_id = 4;
 vec_to_weighted_mean 
----------------------
 {1,NULL,2}
(1 row)

-- real weighted_mean array of nulls and one other
SELECT vec_to_weighted_mean(reals, reals) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_weighted_mean 
----------------------
 {1,NULL,2}
(1 row)

-- float weighted_mean lots
SELECT vec_to_weighted_mean(floats, floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_weighted_mean 
----------------------
 {1,2,2.6}
(1 row)

-- float weighted_mean none
SELECT vec_to_weighted_mean(floats, floats) FROM measurements WHERE sensor_id = -1;
 vec_to_weighted_mean 
----------------------
 
(1 row)

-- float weighted_mean one null
SELECT vec_to_weighted_mean(floats, floats) FROM measurements WHERE sensor_id = 1;
 vec_to_weighted_mean 
----------------------
 
(1 row)

-- float weighted_mean array of nulls
SELECT vec_to_weighted_mean(floats, floats) FROM measurements WHERE sensor_id = 2;
 vec_to_weighted_mean 
----------------------
 {NULL,NULL,NULL}
(1 row)

-- float weighted_mean one not-null
SELECT vec_to_weighted_mean(floats, floats) FROM measurements WHERE sensor_id = 4;
 vec_to_weighted_mean 
----------------------
 {1,NULL,2}
(1 row)

-- float weighted_mean array of nulls and one other
SELECT vec_to_weighted_mean(floats, floats) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_weighted_mean 
----------------------
 {1,NULL,2}
(1 row)

-- numeric weighted mean lots
SELECT vec_to_weighted_mean(nums, nums) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
                    vec_to_weighted_mean                    
------------------------------------------------------------
 {1.2300000000000000,2.3400000000000000,3.0013989637305699}
(1 row)

-- numeric weighted mean none
SELECT vec_to_weighted_mean(nums, nums) FROM measurements WHERE sensor_id = -1;
 vec_to_weighted_mean 
----------------------
 
(1 row)

-- numeric weighted mean one null
SELECT vec_to_weighted_mean(nums, nums) FROM measurements WHERE sensor_id = 1;
 vec_to_weighted_mean 
----------------------
 
(1 row)

-- numeric weighted mean array of nulls
SELECT vec_to_weighted_mean(nums, nums) FROM measurements WHERE sensor_id = 2;
 vec_to_weighted_mean 
----------------------
 {NULL,NULL,NULL}
(1 row)

-- numeric weighted mean one not-null
SELECT vec_to_weighted_mean(nums, nums) FROM measurements WHERE sensor_id = 4;
               vec_to_weighted_mean               
--------------------------------------------------
 {1.23000000000000000000,NULL,2.3400000000000000}
(1 row)

-- numeric weighted mean array of nulls and one other
SELECT vec_to_weighted_mean(nums, nums) FROM measurements WHERE sensor_id IN (2, 4);
               vec_to_weighted_mean               
--------------------------------------------------
 {1.23000000000000000000,NULL,2.3400000000000000}
(1 row)

-- numeric weighted mean everything
SELECT vec_to_weighted_mean(nums, nums) FROM measurements;
                    vec_to_weighted_mean                    
------------------------------------------------------------
 {1.2300000000000000,2.1089847715736041,3.9171756756756757}
(1 row)

-- numeric weighted mean repeating tail
SELECT vec_to_weighted_mean(vals, vals) FROM (VALUES (ARRAY[1]::numeric[]), (ARRAY[0]::numeric[]), (ARRAY[0]::numeric[])) t(vals);
   vec_to_weighted_mean   
--------------------------
 {1.00000000000000000000}
(1 row)

-- numeric weighted mean whole number
SELECT vec_to_weighted_mean(vals, vals) FROM (VALUES (ARRAY[1]::numeric[]), (ARRAY[1]::numeric[])) t(vals);
   vec_to_weighted_mean   
--------------------------
 {1.00000000000000000000}
(1 row)

-- numeric weighted mean data 01
-- results expected to match:
-- WITH di AS (
--   SELECT idx_i, sum(val_i * val_i) / sum(val_i) AS avg_i
--   FROM measurements2 d, unnest(d.data_i) WITH ORDINALITY AS a(val_i, idx_i)
--   GROUP BY idx_i
-- )
-- SELECT array_agg(avg_i ORDER BY idx_i) AS data_i_avg FROM di
SELECT vec_to_weighted_mean(data_i, data_i) FROM measurements2;
                                                                                         vec_to_weighted_mean                                                                                         
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {15512520.937103316864,2393642.708648408277,229.8051872353514445,49.9971640060257455,7405305.981706577723,0.99941945428883988438,-9318.6771749349126032,15924635.570980675290,22237181.258514312241}
(1 row)

